### DISCUSSION TITLE
bam.bio functionality on local files

### DISCUSSION THREAD

**Discussion by RandyHarr** [sandbox-bio/sandbox.bio#11] [Q&A] [answered]
The original bam.bio setup had the ability to use a local BAM file with the webassembly samtools.  This was useful for reading the header, the first few lines, etc.  Or, along with your new expanded tutorial, starting to learn how to use samtools if you are very new to it. 
 Is that functionality to use a local file still available in sandbox.bio?

**Comment by robertaboukhalil** [comment]
Great idea!

**Comment by robertaboukhalil** [comment, marked as answer]
@RandyHarr I've added this functionality in the hamburger menu in the top right corner of the terminal! (you may need to hard reload the page to see it). Let me know what you think!

<img width="461" alt="Screen Shot 2022-05-07 at 9 23 32 PM" src="https://user-images.githubusercontent.com/2009870/167281660-62640871-6201-4c7d-a244-dd8bd8d440f5.png">

**Reply by RandyHarr** [reply]
I think I know the answer (no), but is there a way to save the content output from the tools to the local file system?  (I recall this being a limitation / security issue for browsers to write other than to cookies.)  So only simple output to the terminal that can be copy-paste captured (like from idxstats) is possible to save locally?  That is, you cannot save an index created from a BAM file, correct?  Even if so, having the local files to use is a big help to those trying to begin to learn and understand what their WGS test result returned. The playground is a stepping stone to learning and fully installing the tools to run locally.

**Reply by robertaboukhalil** [reply]
Yes, I was just looking into this! It should be possible. Writing directly to the user's computer with the [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API) isn't supported in all browsers and I'd rather not risk overwriting user data 😅

But, what I'll do is add a `download` command that triggers a local download (even for binary files like BAI indexes), and the user can decide where to save it.

**Reply by robertaboukhalil** [reply]
The `download` command is now ready to be used on sandbox.bio--you might need to clear your cache to get the latest version.

Here's an example usage:

```bash
samtools view -b /samtools/examples/toy.sam -o test.bam
samtools sort test.bam -o test.sorted.bam
samtools index test.sorted.bam
download test.sorted.bam.bai
```

**Reply by RandyHarr** [reply]
Was just trying to use the new Download command again. But cannot recall where it puts the file.  Did not find it in the Users download directory or similar places. Hint?

UPDATE: I see if I use MS Edge, it seems to use the Browsers download function and puts it in the users Downloads folder.  When using Chrome, this does not work. Tried it on multiple installs for Chrome using their latest (105.0.5195.102).

**Reply by robertaboukhalil** [reply]
That's strange, I have Chrome too and it opens a window that asks where I want the file to be saved. Can you double check chrome://downloads just in case?

**Reply by RandyHarr** [reply]
I found the issue; should have noticed before. Privacy Badger is blocking cookies from a site you use: cdn.jsdelivr.net . Once I disable for your site, it works fine on Chrome.
